### PR TITLE
NoMethodError fix for Similar ClassNames and MethodNames

### LINKED ISSUE #24: NoMethodError Missing did_you_mean suggestion when Namespace exists
In many ruby projects, often classes or modules have similar names, which is why did_you_mean comes in handy.  Where it is currently broken, is if you put in the DefinedButWrongNamespace.method.

### Works Correctly

```ruby
class Contact
  def self.first_name
    'bob'
  end

  def self.emoticon
    ":) ;) :P"
  end
end

Contacts.emoticon
#=> NameError: uninitialized constant Contacts
#
#    Did you mean? Contact
```

### Doesn't Work
When you add the plural version of the Contact class, it doesn't work any more.

```ruby
class Contact
  def self.first_name
    'bob'
  end

  def self.emoticon
    ":) ;) :P"
  end
end

class Contacts
  def self.first_names
    %w{bob sue jane}
  end
end

Contacts.emoticon
#=> NoMethodError: undefined method `emoticon' for Contacts:Class
```
Should suggest similar Namespace "Contact"

```ruby
Contacts.emoticon
#=> NoMethodError: undefined method `emoticon' for Contacts:Class
#
#    Did you mean? Contact.emoticon
```

### Another Example

```ruby
Contacts.first_name
#=> NoMethodError: undefined method `first_name' for Contacts:Class
#
#    Did you mean? .first_names
```
Should suggest both alternative method and alternative Namespace.method

```ruby
Contacts.first_name
#=> NoMethodError: undefined method `first_name' for Contacts:Class
#
#    Did you mean? .first_names
#                  Contact.first_name
```

### PR DESCRIPTION
**Pull request by tleish**
(deleted previous pull requests to make it cleaner)

These code changes resolve #24 using the SimilarMethodFinder and SimilarClassFinder.

* Checks to see if the receiver is a Class or Module
* If Class or Module, then use SimilarClassFinder for classes other than current receiver
* For each SimilarClass use SimilarMethodFinder
* Return Results

I also noticed that when in Rails using development mode and cache_class set to false if I have 2 classes (User and Users), and I attempt to access the active record model (e.g. Users.first), the suggestion "Did you mean? User.first" still didn't work unless I had already loaded the User class. I see users do this all the time with active record models, forgetting the plural vs singular form.

To solve this, I added a simple method (if Rails is loaded) that takes the receiver class and attempts to autoload a safe_contantize of the pluralized or singular version of the Class name (see autoload_class_name_inflector). This resolves the issue.

Also improved the order in which suggestions are recommended.

### CODE REVIEW

File: lib/did_you_mean/finders/name_error_finders/similar_class_finder.rb [hunk 1]
```diff
@@ -32,6 +33,13 @@ def scopes
 
     private
 
+    def autoload_class_name_inflector
+      return unless defined?(Rails)
+      name_is_singular = name_from_message.to_s.pluralize.singularize == name_from_message.to_s
+      class_name_inflector = name_is_singular ? name_from_message.to_s.pluralize : name_from_message.to_s.singularize
+      class_name_inflector.safe_constantize
+    end
+
```
Comment (yuki24) [inline lib/did_you_mean/finders/name_error_finders/similar_class_finder.rb]: I don't like this change. Debugging tools should never have side effects. This side effect would confuse people who know how Rails loads constants. If you are using Rails, you should be aware of it.

File: lib/did_you_mean/test_helper.rb [hunk 2]
```diff
@@ -1,7 +1,7 @@
 module DidYouMean
   module TestHelper
     def assert_suggestion(array, expected)
-      assert_equal [expected], array, "Expected #{array.inspect} to only include #{expected.inspect}"
+      assert_equal Array(expected), array, "Expected #{array.inspect} to only include #{expected.inspect}"
```
Comment (yuki24) [inline lib/did_you_mean/test_helper.rb]: If you want to assert multiple suggestions, consider adding `assert_suggestions` method instead of changing simgular `assert_suggestion`. Also, don't include multiple changes in one pull request.

File: lib/did_you_mean/word_collection.rb [hunk 3]
```diff
@@ -1,3 +1,4 @@
+require 'ostruct'
```
Comment (yuki24) [inline lib/did_you_mean/word_collection.rb]: Why is this needed here?

File: lib/did_you_mean/word_collection.rb [hunk 4]
```diff
@@ -14,8 +15,10 @@ def each(&block) words.each(&block); end
     def similar_to(target_word)
       target_word = target_word.to_s
       threshold   = threshold(target_word)
-
-      select {|word| Levenshtein.distance(word.to_s, target_word) <= threshold }
+      map { |word| [Levenshtein.distance(word.to_s, target_word), word] }
+        .select { |distance, _| distance <= threshold }
+        .sort { |a, b| a[0] <=> b[0] }
+        .map { |_, word| word }
```
Comment (yuki24) [inline lib/did_you_mean/word_collection.rb]: I like what you wanted to do here, but creating an array for each word could be slower than several distance re-calculations after `#select`. And again, don't include multiple changes in one pull request.

File: lib/did_you_mean/word_collection.rb (hunk 3 above)
Comment (tleish) [inline lib/did_you_mean/word_collection.rb]: I originally build OpenStructs instead of Arrays, but eventually decided new arrays would be faster than new objects, so I took the OpenStucts out, but forgot to remove the require.

File: lib/did_you_mean/word_collection.rb (hunk 4 above)
Comment (tleish) [inline lib/did_you_mean/word_collection.rb]: Sorry about the multiple include

File: lib/did_you_mean/finders/name_error_finders/similar_class_finder.rb (hunk 1 above)
Comment (tleish) [inline lib/did_you_mean/finders/name_error_finders/similar_class_finder.rb]: If I'm in a Rails console in development mode and type Subscribers.first, then I get "undefined constant" with no suggestion.  I was confused at first as to whydid_you_mean did not suggest the correct Subscriber.first.  However, if I type Subscriber (to load the class) and then type Subscribers.first, I then get the suggestion.

I'm not sure I see the harm in auto-loading a single class when an error occurs to make a single suggestion.  It's only executed for the constant they they mistyped (not all the possible constants available).  And it's only executed when an error occurs and did_you_mean is making a suggestion.

File: lib/did_you_mean/test_helper.rb (hunk 2 above)
Comment (tleish) [inline lib/did_you_mean/test_helper.rb]: Apologies for multiple changes in commit.  Did you want me to make these changes and re-send a pull request?

File: lib/did_you_mean/finders/name_error_finders/similar_class_finder.rb (hunk 1 above)
Comment (yuki24) [inline lib/did_you_mean/finders/name_error_finders/similar_class_finder.rb]: Anyone can override `#const_missing` or do something like `if defined?(Model)` for development purposes. If they assume that constants are auto-loaded by Rails, then `did_you_mean` would break it. In other words, people who assume so will be confused by the side effect. In addition, [Configuring `ActiveSupport::Inflector::Inflections`](http://api.rubyonrails.org/classes/ActiveSupport/Inflector/Inflections.html) could affect the code here.

Also, `#autoload_class_name_inflector ` is too specific as I haven't seen plural/singular class names both together. So I don't think it really helps many people.

File: lib/did_you_mean/test_helper.rb (hunk 2 above)
Comment (yuki24) [inline lib/did_you_mean/test_helper.rb]: Absolutely yes. But I want a separate method that I can use like this:

```ruby
assert_suggestions @exceptions.suggestions, "word1", "word2", ...
```

File: lib/did_you_mean/word_collection.rb (hunk 3 above)
Comment (yuki24) [inline lib/did_you_mean/word_collection.rb]: :+1:

File: lib/did_you_mean/word_collection.rb (hunk 4 above)
Comment (yuki24) [inline lib/did_you_mean/word_collection.rb]: I did a small benchmark and it turns out that this doesn't impact performance and re-calculation of distance is actually slower, so I would like this change to be in.

**Comment by yuki24** [comment]
I'll actually hold off on merging this pull request for now as I wanted to see if this change really helps.

**Comment by tleish** [comment]
No problem.  Let me know if you still want me to break up the pull requests.

File: lib/did_you_mean/word_collection.rb (hunk 4 above)
Comment (yuki24) [inline lib/did_you_mean/word_collection.rb]: I would like this change to be in as a separate pull request. Could you send it?
